File: arrays/0349-intersection-of-two-arrays/java/Solution.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public int[] intersection(int[] nums1, int[] nums2) {
        Arrays.sort(nums1);
        Arrays.sort(nums2);

        List<Integer> intersection = new ArrayList<>();
        int first = 0;
        int second = 0;

        while (first < nums1.length && second < nums2.length) {
            if (nums1[first] == nums2[second]) {
                if (intersection.isEmpty()
                        || intersection.get(intersection.size() - 1) != nums1[first]) {
                    intersection.add(nums1[first]);
                }
                first++;
                second++;
            } else if (nums1[first] < nums2[second]) {
                first++;
            } else {
                second++;
            }
        }

        return intersection.stream().mapToInt(Integer::intValue).toArray();
    }
}

File: arrays/0349-intersection-of-two-arrays/cpp/Solution.cpp
#include <algorithm>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        sort(nums1.begin(), nums1.end());
        sort(nums2.begin(), nums2.end());

        vector<int> result;
        size_t first = 0;
        size_t second = 0;

        while (first < nums1.size() && second < nums2.size()) {
            if (nums1[first] == nums2[second]) {
                if (result.empty() || result.back() != nums1[first]) {
                    result.push_back(nums1[first]);
                }
                first++;
                second++;
            } else if (nums1[first] < nums2[second]) {
                first++;
            } else {
                second++;
            }
        }

        return result;
    }
};
